# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/workers/WorkerRunnable.h
# Commit: b7d4fb0fcc54
# Full Hash: b7d4fb0fcc54c929650660f1c61b63bb4694766f
# Author: Eden Chuang <echuang@mozilla.com>
# Date: 2024-06-25 09:36:02
# Regressor Bug: 1894231
# File Overlap Count: 1
# Description:
#   Bug 1894231 - P11 Remove WorkerThreadRunnable::mWorkerPrivateForPreStartCleaning. r=dom-worker-reviewers,asuth
#   
#   This patch introduces a boolean WorkerThreadRunnable::mCleanPreStartDispatching to indicate if the WorkerThreadRunnable needs to skip its execution in its Run() because of WorkerPrivate::RunLoopNeverRan().
#   
#   If any WorkerPrivate initialization fails in [[ https://searchfox.org/mozilla-central/rev/a26e972e97fbec860400d80df625193f4c88d64e/dom/workers/RuntimeService.cpp#2090-2120 | WorkerThreadPrimaryRunnable::Run() ]], corresponding WorkerJSContext might not be created or created then destroyed before entering [[ https://searchfox.org/mozilla-central/rev/a26e972e97fbec860400d80df625193f4c88d64e/dom/workers/RuntimeService.cpp#2085 | WorkerPrivate::RunLoopNeverRan() ]]. This invalidates the WorkerJSContext, and GetCurrentThreadWorkerPrivate() will not work correctly. In this case, any WorkerThreadRunnable should not run in this invalid environment, so we can just skip the execution by returning NS_OK in its Run().
# ==============================================================================

diff -r bb302697f033 -r b7d4fb0fcc54 dom/workers/WorkerRunnable.h
--- a/dom/workers/WorkerRunnable.h	Mon Jun 24 22:20:54 2024 +0000
+++ b/dom/workers/WorkerRunnable.h	Mon Jun 24 22:20:54 2024 +0000
@@ -264,17 +264,7 @@
   // Cancel(). Only checked and modified on the target thread.
   bool mCallingCancelWithinRun;
 
-  // If dispatching a WorkerThreadRunnable before Worker initialization complete
-  // in worker thread, which are in WorkerPrivate::mPreStartRunnables, when
-  // GetCurrentThreadWorkerPrivate() might get an invalid WorkerPrivate for
-  // WorkerThreadRunnable::Run() because it is in Worker's shutdown.
-  //
-  // This is specific for cleanup these pre-start runnables if the shutdown
-  // starts before Worker executes its event loop.
-  // This member is only set when the runnable is dispatched to
-  // WorkerPrivate::mPreStartRunnables. Any other cases to use this
-  // WorkerPrivate is always wrong.
-  CheckedUnsafePtr<WorkerPrivate> mWorkerPrivateForPreStartCleaning;
+  bool mCleanPreStartDispatching{false};
 };
 
 // This runnable is used to send a message to a worker debugger.
